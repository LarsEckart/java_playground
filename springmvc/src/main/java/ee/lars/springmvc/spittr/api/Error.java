package ee.lars.springmvc.spittr.api;

public class Error {

    private final int code;
    private final String message;

    public Error(int code, String message) {
        this.code = code;
        this.message = message;
    }

    public int getCode() {
        return this.code;
    }

    public String getMessage() {
        return this.message;
    }
}