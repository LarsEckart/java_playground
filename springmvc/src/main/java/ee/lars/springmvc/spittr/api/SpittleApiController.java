package ee.lars.springmvc.spittr.api;

import ee.lars.springmvc.spittr.Spittle;
import ee.lars.springmvc.spittr.data.SpittleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/spittles")
public class SpittleApiController {

    private static final String MAX_LONG_AS_STRING = "9223372036854775807";

    private final SpittleRepository spittleRepository;

    @Autowired
    public SpittleApiController(SpittleRepository spittleRepository) {
        this.spittleRepository = spittleRepository;
    }

    @RequestMapping(method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
    public List<Spittle> spittles(
            @RequestParam(value = "max", defaultValue = MAX_LONG_AS_STRING) long max,
            @RequestParam(value = "count", defaultValue = "20") int count)
    {
        return this.spittleRepository.findSpittles(max, count);
    }

    @RequestMapping(method = RequestMethod.POST, consumes = "application/json")
    public Spittle saveSpittle(@RequestBody Spittle spittle) {
        return spittleRepository.save(spittle);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public ResponseEntity<?> spittleById(@PathVariable long id) {
        Spittle spittle = this.spittleRepository.findOne(id);
        if (spittle == null) {
            Error error = new Error(4, "Spittle [" + id + "] not found");
            return new ResponseEntity<Error>(error, HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<Spittle>(spittle, HttpStatus.OK);
    }
}
