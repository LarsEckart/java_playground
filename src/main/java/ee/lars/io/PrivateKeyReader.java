package ee.lars.io;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.spec.EncodedKeySpec;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;

public class PrivateKeyReader {

    /**
     * Reads a private key file which is in PKCS#8 format.
     * For example create a private key via
     * openssl genrsa -out private_key.pem 2048
     * and then convert it via
     * openssl pkcs8 -topk8 -inform PEM -outform DER -in private_key.pem -out private_key.der -nocrypt
     */
    public static PrivateKey get(String filename)
            throws IOException, NoSuchAlgorithmException, InvalidKeySpecException
    {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));

        EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        final PrivateKey privateKey = keyFactory.generatePrivate(keySpec);
        return privateKey;
    }
}
