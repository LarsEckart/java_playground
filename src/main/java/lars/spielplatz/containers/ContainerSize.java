package lars.spielplatz.containers;

import java.lang.reflect.Field;
import java.util.HashMap;

class ContainerSize {


  private enum States {
    dummy_val_01,
    dummy_val_02,
    dummy_val_03,
    dummy_val_04,
    dummy_val_05,
    dummy_val_06,
    dummy_val_07,
    dummy_val_08,
    dummy_val_09,
    dummy_val_10,
    dummy_val_11,
    dummy_val_12,
    dummy_val_13,
    dummy_val_14,
    dummy_val_15,
    dummy_val_16,
    dummy_val_17,
    dummy_val_18,
    dummy_val_19,
    dummy_val_20,
  }

  /*
   * add "--add-opens java.base/java.util=ALL-UNNAMED" to vm options in intellij otherwise this
   * won't work.
   */
  public static void main(String[] args) throws Exception {
    var oldWay = new HashMap<>(States.values().length);
    var newWay = HashMap.newHashMap(States.values().length);

    System.out.println("threshold after creation");
    System.out.println(getThreshold(newWay));
    System.out.println(getThreshold(oldWay));

    for (States s : States.values()) {
      oldWay.put(s.name(), s);
      newWay.put(s.name(), s);
    }

    System.out.println("threshold after filling");
    System.out.println(getThreshold(newWay));
    System.out.println(getThreshold(oldWay));
  }

  private static int getThreshold(HashMap<Object, Object> map)
      throws NoSuchFieldException, IllegalAccessException {
    Field f = map.getClass().getDeclaredField("threshold"); //NoSuchFieldException
    f.setAccessible(true);
    return (int) f.get(map); //IllegalAccessException
  }

}
