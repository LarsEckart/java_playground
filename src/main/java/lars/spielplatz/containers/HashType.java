package lars.spielplatz.containers;

public class HashType extends SetType {

  public HashType(int i) {
    super(i);
  }

  @Override
  public int hashCode() {
    return i;
  }
}
