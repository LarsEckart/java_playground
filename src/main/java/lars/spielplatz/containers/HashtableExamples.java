package lars.spielplatz.containers;

import java.util.HashMap;
import java.util.Hashtable;
import java.util.Map;
import java.util.concurrent.atomic.LongAdder;

class HashtableExamples {

  static class Person {
    private final String name;

    public Person(String name) {
      this.name = name;
    }

    @Override
    public int hashCode() {
      return 1;
    }

    @Override
    public boolean equals(Object obj) {
      if (!(obj instanceof Person)) return false;
      System.out.println("equals(" + ((Person) obj).name + ")");
      return name.equals(((Person) obj).name);
    }
  }

  public static void main(String[] args) {
    Map<Person, Object> hashtable = new Hashtable<>();
    hashtable.put(new Person("Lars"), "any");
    System.out.println(hashtable);
    System.out.println(hashtable.get(new Person("Oskar")));

    Map<Person, Object> map = new HashMap<>();
    map.put(new Person("Lars"), "any");
    System.out.println(map);
    System.out.println(map.get(new Person("Oskar")));

    LongAdder la = new LongAdder();
    la.increment();
  }
}
