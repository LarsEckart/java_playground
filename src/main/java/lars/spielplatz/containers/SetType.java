package lars.spielplatz.containers;

public class SetType {

  int i;

  public SetType(int i) {
    this.i = i;
  }

  @Override
  public boolean equals(Object o) {
    return o instanceof SetType && (i == ((SetType) o).i);
  }

  @Override
  public String toString() {
    return Integer.toString(i);
  }
}
