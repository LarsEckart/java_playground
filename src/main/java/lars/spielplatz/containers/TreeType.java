package lars.spielplatz.containers;

public class TreeType extends SetType implements Comparable<TreeType> {

    public TreeType(int i) {
        super(i);
    }

    @Override public int compareTo(TreeType arg) {
        return Integer.compare(arg.i, i);
    }
}
