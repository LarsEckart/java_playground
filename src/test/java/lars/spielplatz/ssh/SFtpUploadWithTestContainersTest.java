package lars.spielplatz.ssh;

import static org.slf4j.LoggerFactory.getLogger;

import com.jcraft.jsch.ChannelSftp;
import com.jcraft.jsch.ChannelSftp.LsEntry;
import com.jcraft.jsch.JSch;
import com.jcraft.jsch.JSchException;
import com.jcraft.jsch.Session;
import com.jcraft.jsch.SftpException;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Vector;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.scrubbers.DateScrubber;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.DisabledOnOs;
import org.junit.jupiter.api.condition.OS;
import org.junit.jupiter.api.io.TempDir;
import org.slf4j.Logger;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.images.builder.ImageFromDockerfile;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

@DisabledOnOs(OS.WINDOWS) // no docker on github actions windows
@Testcontainers
public class SFtpUploadWithTestContainersTest {

  private static final Logger log = getLogger(SFtpUploadWithTestContainersTest.class);

  private static final String USERNAME = "user";
  private static final String PASSWORD = "password";
  private static final String REMOTE_PATH = "upload";

  @TempDir
  private static File directory;

  @Container
  private static final GenericContainer<?> sftp = new GenericContainer<>(
      new ImageFromDockerfile()
          .withDockerfileFromBuilder(builder ->
              builder
                  .from("atmoz/sftp:latest")
                  .run("mkdir -p /home/" + USERNAME + "/upload; chmod -R 007 /home/" + USERNAME)
                  .build()))
      .withExposedPorts(22)
      .withCommand(USERNAME + ":" + PASSWORD + ":1001:::" + REMOTE_PATH);

  @Test
  void uploadOnlyCsvFiles() throws IOException {
    var folderWithFiles = createTempFilesOnLocalStorage(
        "users1.csv", "users2.csv", "users3.csv", "secrets.txt", "config", "users4.csv");

    newUploadAllCsvFiles(folderWithFiles);
  }

  private static File createTempFilesOnLocalStorage(String... fileNames) {
    var files = Arrays.asList(fileNames);
    files.forEach(fileName -> {
      File csvFile = new File(directory.getAbsolutePath() + "/" + fileName);
      try {
        csvFile.createNewFile();
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    });
    return new File(directory.getAbsolutePath());
  }

  private void legacyUploadAllCsvFiles(File directory) {
    final JSch jsch = new JSch();
    Session session = null;
    ChannelSftp channel = null;
    try {
      session = jsch.getSession(USERNAME, sftp.getHost(), sftp.getFirstMappedPort());
      session.setPassword(PASSWORD);
      java.util.Properties config = new java.util.Properties();
      config.put("StrictHostKeyChecking", "no");
      session.setConfig(config);
      session.connect();
      channel = (ChannelSftp) session.openChannel("sftp");
      channel.connect();

      String path = "/upload/";
      channel.cd(path);

      for (File file : directory.listFiles()) {
        if (file.isFile()) {
          String extension = "";
          int i = file.getName().lastIndexOf('.');
          if (i > 0) {
            extension = file.getName().substring(i + 1);
          }
          if ("csv".equals(extension)) {
            log.info("Uploading {}", file.getName());
            try (FileInputStream input = new FileInputStream(file)) {
              try {
                //write to ftp server
                channel.put(input, file.getName());
              } catch (SftpException e) {
                log.error("Failed to upload file {}", file.getName(), e);
              }
            }
          }
        }
      }

      Vector<LsEntry> files = channel.ls(path);
      String collect = files.stream().map(LsEntry::toString).sorted()
          .collect(Collectors.joining("\n"));

      Approvals.verify(collect, new Options(new DateScrubber("[A-Za-z]{3} \\d{2} \\d{2}:\\d{2}")));

    } catch (JSchException | SftpException | IOException e) {
      throw new RuntimeException(e);
    } finally {
      if (channel != null) {
        channel.disconnect();
      }
      if (session != null) {
        session.disconnect();
      }
    }
  }

  private void newUploadAllCsvFiles(File directory) {
    try (SftpOperations ssh = SftpOperations.withJsch(USERNAME, PASSWORD, sftp.getHost(),
        sftp.getFirstMappedPort())) {
      Path path = Path.of("/upload/");
      ssh.changeDirectory(path);

      try (Stream<Path> stream = Files.list(directory.toPath())) {
        stream
            .filter(file -> !Files.isDirectory(file))
            .filter(p -> "csv".equals(fileExtension(p.toFile().getName())))
            .forEach(ssh::copyTo);
      } catch (IOException e) {
        throw new RuntimeException(e);
      }

      List<String> files = ssh.listFiles(path).stream().sorted().toList();
      String collect = String.join("\n", files);
      Approvals.verify(collect, new Options(new DateScrubber("[A-Za-z]{3} \\d{2} \\d{2}:\\d{2}")));
    }
  }

  public String fileExtension(String filename) {
    return Optional.ofNullable(filename)
        .filter(f -> f.contains("."))
        .map(f -> f.substring(filename.lastIndexOf(".") + 1))
        .orElse("");
  }
}

