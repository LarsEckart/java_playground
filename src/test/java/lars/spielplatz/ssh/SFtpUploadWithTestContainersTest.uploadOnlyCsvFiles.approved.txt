users1.csv
users2.csv
users3.csv
users4.csv